### PR TITLE
Shroud, ~23% decrease in tick time.

### PR DESCRIPTION
**Pull request by anvilvapre**
Profiling showed that when a logic tick takes 40% of CPU time
    that about 10%+ of time is spent in updating the Shroud. Roughly
    5% in CellLayer Index and and GetItem[MPos].
    
    The shroud used many cell layers. Each tick all projected
    map cells are enumated. MPos per PPos are allocated and
    cell layer cells are looked up by calculacting their index
    in the cell layer entries map.
    
    In this new version we ask the cell layer who is aware of the
    map grid type to provide us with a enumeration of cells over
    a rectangular region of the layer/map.
    
    Advantages:
    - No MPos to PPos conversion on rectangular grid type cell layers/maps.
    - No repeated calculation of cell indexes in the cell layer entry map.
    - No property based cell lookup by MPos in the cell layer.
    - One cell layer in Shroud, less lookups.
    - Close to the original implementation.

    Disadvantage: no access to MPos, so stored as overhead in ShroudCell. 
    Alternative would be to provide a enumerator that returns a pair MPos and
    Cell but that will not benefit performance. A good alternative would be to make
    access to Cell layer entries public and allow cell access based by index.
    
    Limited tested on a - rectangular - replay map. Please validate and test.
    
    Assumptions: grid type rectangular means PPos == MPos. And that Shroud cell layer
    is static and matches the map size.
 

Benchmark results before and after. First two are after. Second two are before/current release. Replay of a 5.12 minute skirmisch game at max speed.

tick time
```      
 tick_time185446  tick_time185402  tick_time185046  tick_time184716
count      9765.000000      9765.000000      9765.000000      9765.000000
mean          2.649446         2.610178         3.339432         3.357306
std           1.213728         1.221905         1.190524         1.178373
min           1.200400         1.024900         1.959700         1.969700
25%           1.704800         1.645200         2.457500         2.477100
50%           2.493100         2.452900         3.077900         3.150700
75%           3.224700         3.229200         3.859400         3.872800
max          20.434700        19.883100        21.090700        22.343700

```

image

image

No (significant) impact on render time. No projected cell region enumeration used there.

### CODE REVIEW

**Comment by anvilvapre** [comment]
Only OpenRA.Game/Map/CellLayer.cs and OpenRA.Game/Traits/Player/Shroud.cs changed in the initial commit.

I've now merged with bleed. Had to make some changes because `Map.ProjectedCellRegion` no longer exists. The `CellLayer` enumerator still takes a `ProjectedCellRegion`, perhaps this should become and array i.e. `Map.ProjectedCells`. See commit comments.

File: OpenRA.Game/Traits/Player/Shroud.cs [hunk 1]
```diff
@@ -233,21 +236,21 @@ public void AddSource(object key, SourceType type, PPos[] projectedCells)
 					continue;
 
 				var uv = (MPos)puv;
-				touched[uv] = true;
+				var cell = layer[uv];
```
Comment (tovl) [inline OpenRA.Game/Traits/Player/Shroud.cs:237]: Shouldn't this be followed by `Cell.Touched = true`?

File: OpenRA.Game/Traits/Player/Shroud.cs [hunk 2]
```diff
@@ -265,17 +268,17 @@ public void RemoveSource(object key)
 				if (map.Contains(puv))
 				{
 					var uv = (MPos)puv;
-					touched[uv] = true;
+					var cell = layer[uv];
```
Comment (tovl) [inline OpenRA.Game/Traits/Player/Shroud.cs:270]: Same here.

**Comment by pchote** [comment]
Can you please rebase rather than merge to properly fix the commits here?

I would appreciate if you could also rerun your tests after rebasing to give a fair comparison now that #18167 has been merged.

**Review by abcdefg30 — COMMENTED** [review]
Why are we casting to `MPos` again instead of using the `ProjectedCellLayer`?

**Comment by anvilvapre** [comment]
I notice in bleed you are now using `ProjectedCellLayer` over `CellLayer`. This solved the MPos to PPos conversion. Also nice that ProjectedCells is an array.

I do still think iterating over over cells rather than cell positions has a large benefit. Or as an alternative to use an integer index as key to a cell layer over a PPos or MPos. The allows you to easily retrieve the PPos again if needed.  It avoids 8000+ x number of players per logic performing calling Index() and getItem. 

I'll close this pull request. Rebase may have gone wrong. Let me know if you are open to the idea of the cell enumerator. If so i may propose a new pull request in future.

Replay doesn't seem to work in bleed currently. I may run a comparison in future.

**Pull request closed without merging**